00 Core/MWSE/mods/tew/AURA/fader.lua

```diff
@@ -2,55 +2,61 @@ local this = {}
 
 local common = require("tew.AURA.common")
 local debugLog = common.debugLog
-local moduleData = require("tew.AURA.moduleData")
+local modules = require("tew.AURA.modules")
+local moduleData = modules.data
 local volumeController = require("tew.AURA.volumeController")
-local adjustVolume = volumeController.adjustVolume
 
 local TICK = 0.1
 local MAX = 1
 local MIN = 0
 
-local function parse(options)
-    local moduleName = options.module
-    local ref = options.reference
-    local track = options.track
+this.inProgress = {["in"] = {}, ["out"] = {}}
+
+local function getInProgressCount(moduleName, fadeType)
+    local count = 0
+    for _, fade in ipairs(this.inProgress[fadeType]) do
+        if (fade.moduleName == moduleName) then
+            count = count + 1
+        end
+    end
+    return count
+end
+
+function this.fade(options)
+    local moduleName = options.module or "n/a"
+    local mData = moduleData[moduleName]
     local fadeType = options.fadeType
-	local oppositeType = (fadeType == "out") and "in" or "out"
-	local lastVolume = moduleData[options.module].lastVolume
-    local volume = options.volume or lastVolume or track.volume
-    local pitch = options.pitch or MAX
-    local targetDuration = options.duration or moduleData[moduleName].faderData[fadeType].duration
-    local currentVolume
-	local fadeInProgress = {}
-	local iterTimer
-	local fadeTimer
-
-
-    if not (track and ref) then
-        debugLog(string.format("[!][%s] Track: %s, ref: %s. Returning.", moduleName, tostring(track), tostring(ref)))
+    local fadeTypeOpposite = (fadeType == "out") and "in" or "out"
+    local track = options.track
+    local trackId = track and track.id
+    local ref = options.reference
+    local removeTrack = options.removeTrack -- Whether to remove the track after fading it out
+    local isTrackUnattached = not ref
+
+    if not (fadeType and track) then
+        debugLog(string.format("[!][%s] Track: %s, fadeType: %s. Returning.", moduleName, tostring(track), tostring(fadeType)))
         return
     end
 
-	if this.isRunning{
+    if this.isRunning{
 		module = moduleName,
 		fadeType = fadeType,
 		track = track,
 		reference = ref,
 	} then
-		--debugLog(string.format("[%s] Already fading %s %s on %s. Returning.", moduleName, fadeType, track.id, tostring(ref)))
 		return
 	end
 
-	if this.isRunning{
+    if this.isRunning{
 		module = moduleName,
-		fadeType = oppositeType,
+		fadeType = fadeTypeOpposite,
 		track = track,
 		reference = ref,
 	} then
-        debugLog(string.format("[%s] wants to fade %s %s but fade %s is in progress for this track. Trying later.", moduleName, fadeType, track.id, oppositeType))
+        debugLog(string.format("[%s] wants to fade %s %s but fade %s is in progress for this track. Trying later.", moduleName, fadeType, trackId, fadeTypeOpposite))
```

Comment:
There's a special place in heaven for people who bother doing this kind of stuff, brilliant :clap: